svf/lib/SABER/SaberCondAllocator.cpp

```diff
@@ -393,10 +393,7 @@ bool SaberCondAllocator::isTestContainsNullAndTheValue(const CmpStmt *cmp) const
         Set<const SVFValue* > inDirVal;
         for (const auto &it: getCurEvalSVFGNode()->getOutEdges())
         {
-            if (it->isIndirectVFGEdge())
-            {
-                inDirVal.insert(it->getDstNode()->getValue());
-            }
+            inDirVal.insert(it->getDstNode()->getValue());
```

Comment:
Can you add the description for the non-indirect value-flow case at/from line 385 and modify the comments where possible?

Reply:
So I will modify the head comment for function `isTestContainsNullAndTheValue`, is that ok?

Reply:
Should we add the value of the current SVFG Node (%3 in this example) if it is directly passed as a CMP operand? I think `inDirVal.insert(getCurEvalSVFGNode()->getValue());` should be inserted before iterating the out edges of the current VFG node.

Reply:
@seviezhou can you have a look at @jumormt's comment?

Reply:
From my debug result, both `op0` and `getCurEvalSVFGNode` return `%3`, and it has 2 out edges, both of them also return `%3`:

```
op0:    %3 = tail call i8* @malloc(i64 16), !dbg !22 ln: 10 fl: part-leak.c
getCurEvalSVFGNode: AddrVFGNode ID: 5 AddrStmt: [Var22 <-- Var23]
   %3 = tail call i8* @malloc(i64 16), !dbg !22 ln: 10 fl: part-leak.c

num edge: 2
dstNode: ActualParmVFGNode ID: 9 CS[ln: 13 fl: part-leak.c]ValVar ID: 22
   %3 = tail call i8* @malloc(i64 16), !dbg !22 ln: 10 fl: part-leak.c
insert dstNode:    %3 = tail call i8* @malloc(i64 16), !dbg !22 ln: 10 fl: part-leak.c

dstNode: ActualParmVFGNode ID: 21 CS[ln: 13 fl: part-leak.c]ValVar ID: 22
   %3 = tail call i8* @malloc(i64 16), !dbg !22 ln: 10 fl: part-leak.c
insert dstNode:    %3 = tail call i8* @malloc(i64 16), !dbg !22 ln: 10 fl: part-leak.c

inDirVal.find(op0) != inDirVal.end(): true
```

I am not very familiar with the internal of SVF, so I am not sure how the out edges are constructed. But based on this results, maybe we do not need to add `inDirVal.insert(getCurEvalSVFGNode()->getValue())`, but for insurance, we can add it.

Reply:
@seviezhou Thanks for your detailed answer. It is correct that these SVFG nodes contain the same variable (both `%3`) because the direct value flow is a top-level variable in the LLVM partial SSA form. I think it is necessary to add this line in case there are no direct out edges of the current SVFG node.

Reply:
Ok, I will make a commit.

Reply:
Patch like this would be ok:

```diff
@@ -391,6 +391,7 @@ bool SaberCondAllocator::isTestContainsNullAndTheValue(const CmpStmt *cmp) const
    if (SVFUtil::isa<SVFConstantNullPtr>(op1))
    {
        Set<const SVFValue* > inDirVal;
+       inDirVal.insert(getCurEvalSVFGNode()->getValue());
        for (const auto &it: getCurEvalSVFGNode()->getOutEdges())
        {
            if (it->isIndirectVFGEdge())
@@ -403,6 +404,7 @@ bool SaberCondAllocator::isTestContainsNullAndTheValue(const CmpStmt *cmp) const
    else if (SVFUtil::isa<SVFConstantNullPtr>(op0))
    {
        Set<const SVFValue* > inDirVal;
+       inDirVal.insert(getCurEvalSVFGNode()->getValue());
        for (const auto &it: getCurEvalSVFGNode()->getOutEdges())
        {
            if (it->isIndirectVFGEdge())
```